File: FE/app/src/test/java/com/ssafy/e102/eumgil/app/navigation/AppStartDestinationTest.java
package com.ssafy.e102.eumgil.app.navigation;

import static org.junit.Assert.assertEquals;

import com.ssafy.e102.eumgil.core.model.AuthGateState;
import com.ssafy.e102.eumgil.core.model.AuthSession;
import com.ssafy.e102.eumgil.core.model.InitSettings;
import com.ssafy.e102.eumgil.feature.onboarding.PrimaryUserType;
import org.junit.Test;

public class AppStartDestinationTest {
    @Test
    public void sessionMissingStartsAtLoginRegardlessOfCompletedOnboarding() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(null, false),
                        completedInitSettings());

        assertEquals(AuthRoute.Login.INSTANCE.getRoute(), destination.getRoute());
    }

    @Test
    public void profileIncompleteSessionStartsAtProfileSetup() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(authSession(), false),
                        completedInitSettings());

        assertEquals(AuthRoute.ProfileSetup.INSTANCE.getRoute(), destination.getRoute());
    }

    @Test
    public void profileCompletedSessionStartsOnboardingWhenOnboardingIsMissing() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(authSession(), true),
                        new InitSettings(null, null, false, false, false));

        assertEquals(OnboardingRoute.UserTypePrimary.INSTANCE.getRoute(), destination.getRoute());
    }

    @Test
    public void profileCompletedSessionReturnsToOnb001WhenOnlyPrimaryUserTypeExists() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(authSession(), true),
                        new InitSettings(
                                PrimaryUserType.LOW_VISION.getRouteValue(),
                                null,
                                false,
                                false,
                                false));

        assertEquals(OnboardingRoute.UserTypePrimary.INSTANCE.getRoute(), destination.getRoute());
    }

    @Test
    public void completedLowVisionSessionStartsAtLowVisionHome() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(authSession(), true),
                        new InitSettings(
                                PrimaryUserType.LOW_VISION.getRouteValue(),
                                null,
                                true,
                                true,
                                true));

        assertEquals("low_vision_graph", destination.getRoute());
    }

    @Test
    public void completedSessionProfileAndOnboardingStartsAtMap() {
        AppStartDestination destination =
                AppStartDestinationKt.resolveAppStartDestination(
                        new AuthGateState(authSession(), true),
                        completedInitSettings());

        assertEquals(TopLevelRoute.Map.INSTANCE.getRoute(), destination.getRoute());
    }

    private static AuthSession authSession() {
        return new AuthSession("access-token", "refresh-token");
    }

    private static InitSettings completedInitSettings() {
        return new InitSettings(
                PrimaryUserType.MOBILITY_IMPAIRED.getRouteValue(),
                "manual_wheelchair",
                false,
                true,
                true);
    }
}

File: FE/app/src/test/java/com/ssafy/e102/eumgil/app/navigation/AppStartDestinationTest.kt
package com.ssafy.e102.eumgil.app.navigation

import com.ssafy.e102.eumgil.core.model.AuthGateState
import com.ssafy.e102.eumgil.core.model.AuthSession
import com.ssafy.e102.eumgil.core.model.InitSettings
import com.ssafy.e102.eumgil.feature.onboarding.PrimaryUserType
import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Test

class AppStartDestinationRoutingTest {
    @Test
    fun `unauthenticated session starts at login route`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(),
                initSettings = completedInitSettings,
            )

        assertSame(AppStartDestination.Login, destination)
        assertEquals(AuthRoute.Login.route, destination.route)
    }

    @Test
    fun `pending signup token starts at onboarding primary user type route`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(signupToken = "signup-token"),
                initSettings = InitSettings(),
            )

        assertSame(AppStartDestination.UserTypePrimaryStep, destination)
        assertEquals(OnboardingRoute.UserTypePrimary.route, destination.route)
    }

    @Test
    fun `authenticated profile incomplete session starts at AUTH-002 route`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(
                    authSession = AuthSession(accessToken = "test-token"),
                    isProfileCompleted = false,
                ),
                initSettings = completedInitSettings,
            )

        assertSame(AppStartDestination.ProfileSetup, destination)
        assertEquals("auth/profile_setup", destination.route)
    }

    @Test
    fun `profile complete session keeps existing onboarding gate`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(
                    authSession = AuthSession(accessToken = "test-token"),
                    isProfileCompleted = true,
                ),
                initSettings = InitSettings(),
            )

        assertSame(AppStartDestination.UserTypePrimaryStep, destination)
    }

    @Test
    fun `profile complete low vision session can force terms guide for local debug`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(
                    authSession = AuthSession(accessToken = "test-token"),
                    isProfileCompleted = true,
                ),
                initSettings =
                    InitSettings(
                        selectedPrimaryUserType = PrimaryUserType.LOW_VISION.routeValue,
                        isLowVisionFollowUpCompleted = true,
                        isLocationTermsAgreed = true,
                        isPrivacyPolicyAgreed = true,
                    ),
                forceLowVisionTermsGuide = true,
            )

        assertSame(AppStartDestination.LowVisionTermsGuide, destination)
        assertEquals(OnboardingRoute.TermsGuide.createRoute(), destination.route)
    }

    @Test
    fun `local debug low vision terms guide keeps login gate`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(),
                initSettings = completedInitSettings,
                forceLowVisionTermsGuide = true,
            )

        assertSame(AppStartDestination.Login, destination)
    }

    @Test
    fun `profile complete session still starts ONB-001 when only primary user type is stored`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(
                    authSession = AuthSession(accessToken = "test-token"),
                    isProfileCompleted = true,
                ),
                initSettings =
                    InitSettings(
                        selectedPrimaryUserType = PrimaryUserType.LOW_VISION.routeValue,
                    ),
            )

        assertSame(AppStartDestination.UserTypePrimaryStep, destination)
    }

    @Test
    fun `completed low vision session starts at low vision home`() {
        val destination =
            resolveAppStartDestination(
                authGateState = AuthGateState(
                    authSession = AuthSession(accessToken = "test-token"),
                    isProfileCompleted = true,
                ),
                initSettings =
                    InitSettings(
                        selectedPrimaryUserType = PrimaryUserType.LOW_VISION.routeValue,
                        isLowVisionFollowUpCompleted = true,
                        isLocationTermsAgreed = true,
                        isPrivacyPolicyAgreed = true,
                    ),
        )

        assertSame(AppStartDestination.LowVisionHome, destination)
        assertEquals(LOW_VISION_GRAPH_ROUTE, destination.route)
    }

    private companion object {
        val completedInitSettings =
            InitSettings(
                selectedPrimaryUserType = PrimaryUserType.MOBILITY_IMPAIRED.routeValue,
                selectedMobilitySubtype = "manual_wheelchair",
                isLocationTermsAgreed = true,
            )
    }
}
